<a href="https://colab.research.google.com/github/Thitiphorn267-3/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [26]:
import pandas as pd
import os

In [27]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/bsc_dpdm25_data'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [28]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/bsc_dpdm25_data/reduced_marketbasket.csv'),encoding='latin-1')

In [29]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

Quiz
มีกี่ประเทศ

In [30]:
set

set

#แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย groupby .apply(list)


In [31]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [32]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [33]:
!pip install apyori

In [34]:
from apyori import apriori

In [35]:
list(apriori(transactions,min_support = 0.065))

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [36]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [37]:
len(frequentItemsets)

4

In [38]:
frequentItemsets[0]# ไอเทมเซตแรก

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [39]:
frequentItemsets[0][0]

frozenset({'22423'})

In [40]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [41]:
frequentItemsets[0][0].__len__()

1

In [42]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [43]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

#Quiz 22386 และ 85099B คือสินค้าอะไร

In [44]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B')]['Description'])

{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}

#HW1

*  หา Frequent Item Set และ Association Rules
* ของช่วงเวลา Chrhistmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลาตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
* ของช่วงเวลา Chrhistmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ + เดือน (groupby ประเทศกับเดือน)

In [45]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/bsc_dpdm25_data/reduced_marketbasket.csv'),encoding='latin-1')
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [46]:
import pandas as pd
import os

In [47]:
!pip install apyori

from apyori import apriori

In [48]:
#ข้อมูล กี่ประเทศ
data.groupby(['Country']).sum()


<ipython-input-48-417e0038914b>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  data.groupby(['Country']).sum()


Quantity    UnitPrice    CustomerID
Country                                                  
Australia                83653     4054.750  1.569300e+07
Austria                   4827     1701.520  5.021102e+06
Bahrain                    260       86.570  2.100270e+05
Belgium                  23152     7540.130  2.571829e+07
Brazil                     356      142.600  4.086080e+05
Canada                    2763      910.580  2.615483e+06
Channel Islands           9479     3738.550  1.128522e+07
Cyprus                    6317     3920.070  7.715880e+06
Czech Republic             592       88.150  3.834300e+05
Denmark                   8188     1266.950  4.876734e+06
EIRE                    142637    48447.190  1.103917e+08
European Community         497      294.050  9.215880e+05
Finland                  10666     3786.850  8.699324e+06
France                  110480    43031.990  1.076489e+08
Germany                 117448    37666.000  1.200751e+08
Greece                    1556      713.290  2.008584e+06
Hong Kong                 4769    12241.500  0.000000e+00
Iceland                   2458      481.210  2.247154e+06
Israel                    4353     1079.040  3.164467e+06
Italy                     7999     3879.390  1.015666e+07
Japan                    25218      814.860  4.567292e+06
Lebanon                    386      242.440  5.743800e+05
Lithuania                  652       99.440  5.366200e+05
Malta                      944      666.010  2.158496e+06
Netherlands             200128     6492.550  3.419054e+07
Norway                   19247     6529.060  1.350765e+07
Poland                    3653     1422.270  4.341972e+06
Portugal                 16180    13037.540  1.886480e+07
RSA                        352      248.100  7.218680e+05
Saudi Arabia                75       24.110  1.256500e+05
Singapore                 5234    25108.890  2.918376e+06
Spain                    26824    12633.450  3.268929e+07
Sweden                   35637     1806.830  6.790083e+06
Switzerland              30325     6813.690  2.377592e+07
USA                       1034      644.980  3.672086e+06
United Arab Emirates       982      229.890  1.018952e+06
United Kingdom         4263829  2245715.474  5.626433e+09
Unspecified               3300     1204.010  3.348046e+06

In [60]:

# แปลงรูปแบบเวลาใน  InvoiceDate เป็น  datetime
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

#เแปลงรูปแบบให้เป็น เดือน / วัน / ปี / ชม. / นท.
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# สร้างคอลัมน์ 'Date' ดึงข้อมูลเฉพาะ
data['Date'] = data['InvoiceDate'].dt.date
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time

data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
...                    ...        ...         ...             ...         ...   
541904 2011-12-09 12:50:00       0.85     12680.0          France  2011-12-09   
541905 2011-12-09 12:50:00       2.10     12680.0          France  2011-12-09   
541906 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541907 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541908 2011-12-09 12:50:00       4.95     12680.0          France  2011-12-09   

        Month  Year      Time  
0          12  2010  08:26:00  
1          12  2010  08:26:00  
2          12  2010  08:26:00  
3          12  2010  08:26:00  
4          12  2010  08:26:00  
...       ...   ...       ...  
541904     12  2011  12:50:00  
541905     12  2011  12:50:00  
541906     12  2011  12:50:00  
541907     12  2011  12:50:00  
541908     12  2011  12:50:00  

[541909 rows x 12 columns]

###1. ข้อมูลของช่วงเวลา Chrhistmas & New Year (1 Dec - 31 Dec)

In [50]:
# ดูช่วงเวลาคริสต์มาสกับปีใหม่
xmas_ny = data[data['Month']==12]
xmas_ny

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
...                    ...        ...         ...             ...         ...   
541904 2011-12-09 12:50:00       0.85     12680.0          France  2011-12-09   
541905 2011-12-09 12:50:00       2.10     12680.0          France  2011-12-09   
541906 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541907 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541908 2011-12-09 12:50:00       4.95     12680.0          France  2011-12-09   

        Month  Year      Time  
0          12  2010  08:26:00  
1          12  2010  08:26:00  
2          12  2010  08:26:00  
3          12  2010  08:26:00  
4          12  2010  08:26:00  
...       ...   ...       ...  
541904     12  2011  12:50:00  
541905     12  2011  12:50:00  
541906     12  2011  12:50:00  
541907     12  2011  12:50:00  
541908     12  2011  12:50:00  

[68006 rows x 12 columns]

In [51]:
trans_xny= xmas_ny.groupby('InvoiceNo')['Description'].apply(list).reset_index()
#
trans_xny

InvoiceNo                                        Description
0       536365  [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       536366  [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       536367  [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       536368  [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4       536369                         [BATH BUILDING BLOCK WORD]
...        ...                                                ...
3035   C581484                      [PAPER CRAFT , LITTLE BIRDIE]
3036   C581490  [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3037   C581499                                           [Manual]
3038   C581568                       [VICTORIAN SEWING BOX LARGE]
3039   C581569  [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...

[3040 rows x 2 columns]

In [52]:
PP= trans_xny['Description'].apply(lambda x: [str(item) for item in x])
PP

0       [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4                              [BATH BUILDING BLOCK WORD]
                              ...                        
3035                        [PAPER CRAFT , LITTLE BIRDIE]
3036    [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3037                                             [Manual]
3038                         [VICTORIAN SEWING BOX LARGE]
3039    [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...
Name: Description, Length: 3040, dtype: object

In [53]:
##กำหนดค่า min_sup = 0.08

#######ไม่ได้ ปรับถี่ใหม่#########


frequentItem_a = list(apriori(PP,min_support = 0.08))
frequentItem_a

[RelationRecord(items=frozenset({"PAPER CHAIN KIT 50'S CHRISTMAS "}), support=0.09243421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({"PAPER CHAIN KIT 50'S CHRISTMAS "}), confidence=0.09243421052631579, lift=1.0)]),
 RelationRecord(items=frozenset({'REGENCY CAKESTAND 3 TIER'}), support=0.08190789473684211, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'REGENCY CAKESTAND 3 TIER'}), confidence=0.08190789473684211, lift=1.0)]),
 RelationRecord(items=frozenset({'WHITE HANGING HEART T-LIGHT HOLDER'}), support=0.09769736842105263, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'WHITE HANGING HEART T-LIGHT HOLDER'}), confidence=0.09769736842105263, lift=1.0)])]

In [54]:
##กำหนดค่า min_sup = 0.035

frequentItem_XN = list(apriori(PP,min_support = 0.035 ))
for i in frequentItem_XN:
    if i[0].__len__()>1:
        print('yes')
        print(i)
#### ตอบบบบ 3 อันก้ได้


yes
RelationRecord(items=frozenset({'ALARM CLOCK BAKELIKE GREEN', 'ALARM CLOCK BAKELIKE RED '}), support=0.03684210526315789, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'ALARM CLOCK BAKELIKE GREEN', 'ALARM CLOCK BAKELIKE RED '}), confidence=0.03684210526315789, lift=1.0), OrderedStatistic(items_base=frozenset({'ALARM CLOCK BAKELIKE GREEN'}), items_add=frozenset({'ALARM CLOCK BAKELIKE RED '}), confidence=0.7088607594936708, lift=14.462662475575565), OrderedStatistic(items_base=frozenset({'ALARM CLOCK BAKELIKE RED '}), items_add=frozenset({'ALARM CLOCK BAKELIKE GREEN'}), confidence=0.7516778523489932, lift=14.462662475575565)])
yes
RelationRecord(items=frozenset({'HOT WATER BOTTLE I AM SO POORLY', 'CHOCOLATE HOT WATER BOTTLE'}), support=0.036513157894736845, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HOT WATER BOTTLE I AM SO POORLY', 'CHOCOLATE HOT WATER BOTTLE'}), confidence=0.036513157894736845, lift=1.0), O

In [55]:
##กำหนดค่า min_sup = 0.035

frequentItem_XN = list(apriori(PP,min_support = 0.035 ))
for i in frequentItem_XN:
    if i[0].__len__()>1:
        print('yes')
        print(i)
#### ตอบบบบ




yes
RelationRecord(items=frozenset({'ALARM CLOCK BAKELIKE GREEN', 'ALARM CLOCK BAKELIKE RED '}), support=0.03684210526315789, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'ALARM CLOCK BAKELIKE GREEN', 'ALARM CLOCK BAKELIKE RED '}), confidence=0.03684210526315789, lift=1.0), OrderedStatistic(items_base=frozenset({'ALARM CLOCK BAKELIKE GREEN'}), items_add=frozenset({'ALARM CLOCK BAKELIKE RED '}), confidence=0.7088607594936708, lift=14.462662475575565), OrderedStatistic(items_base=frozenset({'ALARM CLOCK BAKELIKE RED '}), items_add=frozenset({'ALARM CLOCK BAKELIKE GREEN'}), confidence=0.7516778523489932, lift=14.462662475575565)])
yes
RelationRecord(items=frozenset({'HOT WATER BOTTLE I AM SO POORLY', 'CHOCOLATE HOT WATER BOTTLE'}), support=0.036513157894736845, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HOT WATER BOTTLE I AM SO POORLY', 'CHOCOLATE HOT WATER BOTTLE'}), confidence=0.036513157894736845, lift=1.0), O

*** ตอบ เลือกค่า Suport ที่มากที่สุด คือ
"PAPER CHAIN KIT 50'S CHRISTMAS "และ 'PAPER CHAIN KIT VINTAGE
มีconfidence(ค่าสนับสนุน)=0.0477 หรือ 4.77%

- คนซื้อ PAPER CHAIN KIT 50'S CHRISTMAS และก็จะซื้อ PAPER CHAIN KIT VINTAGE CHRISTMAS มี 51.6%
- คนที่ซื้อ PAPER CHAIN KIT VINTAGE CHRISTMAS แล้วจะซื้อ PAPER CHAIN KIT 50'S CHRISTMAS มี 66.51%



###2. ของประเทศยุโรป

In [56]:
europe_countries = data[(data['Country']!='Austria') | (data['Country']!='Belgium')| (data['Country']!='Channel Islands') | (data['Country']!='Cyprus') | (data['Country']!='Czech Republic') | (data['Country']!='Denmark') |
           (data['Country']!='EIRE') | (data['Country']!='European Community')| (data['Country']!='Finland')| (data['Country']!='France') | (data['Country']!='Germany')| (data['Country']!='Greece')|
           (data['Country']!='Iceland') | (data['Country']!='Italy') | (data['Country']!='Lithuania') | (data['Country']!='Malta') | (data['Country']!='Netherlands') | (data['Country']!='Norway') |
           (data['Country']!='Poland') | (data['Country']!='Portugal') | (data['Country']!='Spain') | (data['Country']!='Sweden') | (data['Country']!='Switzerland') | (data['Country']!='United Kingdom')]
europe_countries #แบ่งทวีปยุโรป

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
...                    ...        ...         ...             ...         ...   
541904 2011-12-09 12:50:00       0.85     12680.0          France  2011-12-09   
541905 2011-12-09 12:50:00       2.10     12680.0          France  2011-12-09   
541906 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541907 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541908 2011-12-09 12:50:00       4.95     12680.0          France  2011-12-09   

        Month  Year      Time  
0          12  2010  08:26:00  
1          12  2010  08:26:00  
2          12  2010  08:26:00  
3          12  2010  08:26:00  
4          12  2010  08:26:00  
...       ...   ...       ...  
541904     12  2011  12:50:00  
541905     12  2011  12:50:00  
541906     12  2011  12:50:00  
541907     12  2011  12:50:00  
541908     12  2011  12:50:00  

[541909 rows x 12 columns]

In [57]:
a = europe_countries.groupby('InvoiceNo')['Description'].apply(list).reset_index()
trans_europe = a['Description'].apply(lambda x: [str(item) for item in x])

In [58]:
trans_europe

0        [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1        [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2        [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3        [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4                               [BATH BUILDING BLOCK WORD]
                               ...                        
25895                        [PAPER CRAFT , LITTLE BIRDIE]
25896    [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
25897                                             [Manual]
25898                         [VICTORIAN SEWING BOX LARGE]
25899    [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...
Name: Description, Length: 25900, dtype: object

In [59]:


frequentItemsets_Europe = list(apriori(trans_Europe, min_support = 0.025))
for i in frequentItemsets_Europe:
    if i[0].__len__()>1:
        print('yes')
        print(i)

        ### ตอบบบ

NameError: name 'trans_Europe' is not defined

*** ค่า Suport ที่มากที่สุด คือ
jumbo bag red retrospot และ jumbo bag pink polkadot
มีconfidence(ค่าสนับสนุน)= 3.22%

- คนที่ซื้อ jumbo bag pink จะซื้อ jumbo bag red มี 67.67%
- คนที่ซื้อjumbo bag red จะซื้อ  jumbo bag pinkS มี 39.02%



### 3. ของช่วงเวลาตรุษจีนในทวีป Asia ( 15 Jan - 15 Feb)

In [ ]:
Asia = ['Hong Kong','Bahrain','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

In [ ]:
ASIA_data = data[data['Country']==""]
for i in range(len(Asia)):
  x = data[data['Country']==Asia[i]]
  ASIA_data=pd.concat([x,ASIA_data])
ASIA_data

In [ ]:
As_data=ASIA_data[(ASIA_data['Month']==1)]
As_data.head()

In [ ]:
As_data_d=As_data[(As_data['Day']>=15)]
As_data_d.head()

In [ ]:
As_data2=ASIA_data[(ASIA_data['Month']==2)]
As_data2.head()

In [ ]:
As_data_d2=As_data2[(As_data2['Day']<=15)]
As_data_d2.head()

In [ ]:
# รวมทั้ง 2 เดือนเข้า ในช่วงตรุษจีน
Asia_data_chi=pd.concat([As_data_d,As_data_d2])
Asia_data_chi

In [ ]:
chi= AS.groupby('InvoiceNo')['Description'].apply(list).reset_index()
chi

In [ ]:
min_sup = 0.082
frequentItemsets_Asia= list(apriori(transactionsAsiaChi,min_support = min_sup))
for i in frequentItemsets_Asia:
    if i[0].__len__()>3:
        print('yes')
        print(i)
frequentItemsets_Asia

# ข้อ 4 ของช่วงเวลา Chrhistmas & New Year ในทวีป Europe (1 Dec - 31 Dec)

In [ ]:
Europe = [
    'Austria', 'Belgium', 'Channel Islands','Cyprus', 'Czech Republic', 'Denmark', 'EIRE',
    'European Community', 'Finland', 'France', 'Germany', 'Greece', 'Iceland', 'Italy',
    'Lithuania', 'Malta', 'Netherlands', 'Norway', 'Poland', 'Portugal', 'Spain', 'Sweden',
    'Switzerland', 'United Kingdom'
]

In [ ]:
EU_data = data[data['Country']==""]
for i in range(len(Europe)):
  x = data[data['Country']==Europe[i]]
  EU_data=pd.concat([x,EU_data])
EU_data

In [ ]:
Eu_data=EU_data[(EU_data['Month']==12)]
Eu_data.head()

In [ ]:
Eu_data_d=Eu_data[(Eu_data['Day']<=31)]
Eu_data_d.head()

In [ ]:
transactionsEU= Eu_data_d.groupby('InvoiceNo')['Description'].apply(list).reset_index()
#
transactionsEU

In [ ]:
min_sup =0.08
frequentItemsetsEU= list(apriori(transactionsEU_C_N,min_support = min_sup))
for i in frequentItemsetsEU:
    if i[0].__len__()>3:
        print('yes')
        print(i)
frequentItemsetsEU

In [ ]:
set(data[(data['StockCode']=='85123A') | (data['StockCode']=='22086')]['Description'])

### 5. Valentine's day


In [ ]:
Valen_data = data[(data['Month']==2)]
Valen_data

In [ ]:
Valen_data = data[(data['Month']==2)]
Valen_data

In [ ]:
transactions_valen = valen_day.groupby('InvoiceNo')['StockCode'].apply(list)
len(transactions_valen)

In [ ]:
min_sup = 0.09
frequentItemsets_valen= list(apriori(transactions_valen,min_support = min_sup))
for i in frequentItemsets_valen:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_valen

In [ ]:
len(frequentItemsets_valen)

In [ ]:
set(data[(data['StockCode']=='20725')|(data['StockCode']=='22720')]['Description'])

- SET OF 3 CAKE TINS PANTRY DESIGN  มีค่า 14.5 % มีมากที่สุดจากทั้งหมด 18 อัน ที่ค่า min sup 0.08

###6. โดยจัดกลุ่มตามประเทศ groupby ประเทศ - เดือน

In [ ]:
data

In [ ]:
trans_CM = data.groupby(['Country','Month'])['Description'].apply(list).reset_index()
trans_Country_M = trans_CM['Description'].apply(lambda x: [str(item) for item in x])
trans_Country_M

In [ ]:
min_sup = 0.31

frequentItemset_CM = list(apriori(trans_Country_M, min_support = min_sup))
for i in frequentItemset_CM:
    if i[0].__len__()>1:
        print('yes')
        print(i)

- มี 2 itemset ที่ผ่าน 0.31
1. คนที่ซื้อทั้ง POSTAGE และ ROUND SNACK BOXES SET OF4 WOODLAND จาก dataset ทั้งหมด มี 32.09%

   คนที่ซื้อ POSTAGE และ ROUND SNACK BOXES SET OF4 WOODLAN มี 32.09%
  
   คนที่ซื้อ POSTAGE แล้วค่อยซื้อ ROUND SNACK BOXES SET OF4 WOODLAND มี 55.88%

   คนที่ซื้อ ROUND SNACK BOXES SET OF4 WOODLAND แล้วค่อยซื้อ POSTAGE มี 84.07%